File: src/app/services/solution-product.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { SolutionProduct } from '../models/solution-product.model';
import { RiderDetailsService } from './rider-details.service';

@Injectable({
  providedIn: 'root'
})
export class SolutionProductService {
  private apiUrl = 'http://localhost:8080/api/solution-product';

  constructor(private http: HttpClient, private riderDetailsService: RiderDetailsService) {}

  manageSolutionProductDetails(solutionProductDetails: SolutionProduct): Observable<SolutionProduct> {
    return this.http.post<SolutionProduct>(`${this.apiUrl}/manage`, solutionProductDetails);
  }

  fetchControlFlagStatus(): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/control-flag-status`);
  }

  updatePremium(premium: number): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/update-premium`, { premium });
  }

  fetchPackages(productCode: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/packages`, { params: { productCode } });
  }

  getPackageList(productCode: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/packages`, { params: { productCode } });
  }

  clearBlocks(): void {
    // Logic to clear the blocks AZBJ_FUND_DETAILS and AZBJ_SSO_FUND_DETAILS
  }

  fetchPayoutFrequency(): Observable<string> {
    return this.http.get<string>(`${this.apiUrl}/payout-frequency`);
  }

  validateFSCCode(fscCode: string): boolean {
    const validFscCodes = ['2000003024', '7000005464', '7000002372', '7000000702'];
    return validFscCodes.includes(fscCode);
  }

  fetchSumAssured(): Observable<number> {
    return this.http.get<number>(`${this.apiUrl}/sum-assured`);
  }

  setPremiumTerm(premiumTerm: number): void {
    this.riderDetailsService.setPremiumTerm(premiumTerm);
  }

  getRenewalPaymentMethods(frequency: string, fscCode: string): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/renewal-payment-methods`, { params: { frequency, fscCode } });
  }

  setBenefitTerm(benefitTerm: number): void {
    // Logic to store the benefit term value
  }

  getBenefitTerm(): number {
    // Logic to retrieve the stored benefit term value
    return 0; // Placeholder return value
  }

  navigateToRiderDetails(): void {
    // Logic to navigate to the rider details block
  }

  calculateAndUpdateRiderCoverAmounts(): void {
    // Logic to calculate and update rider cover amounts
  }

  fetchPaymentMethods(frequency: string, fscCode: string): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/payment-methods`, { params: { frequency, fscCode } });
  }
}

File: src/main/java/com/azbj/proposalde3/service/SolutionProductService.java
package com.azbj.proposalde3.service;

import com.azbj.proposalde3.repository.SolutionProductRepository;
import com.azbj.proposalde3.model.SolutionProduct;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class SolutionProductService {

    @Autowired
    private SolutionProductRepository solutionProductRepository;

    @Autowired
    private FundDetailsService fundDetailsService;

    public SolutionProduct manageSolutionProductDetails(SolutionProduct solutionProductDetails) {
        // Business logic for managing solution product details
        // This is a placeholder for the actual logic
        return solutionProductDetails;
    }

    public List<SolutionProduct> fetchPackages(String productCode) {
        return solutionProductRepository.findByProductCode(productCode);
    }

    public List<SolutionProduct> fetchPackagesByProductId(String productCode) {
        return solutionProductRepository.findByProductId(productCode);
    }

    public List<String> fetchPayoutFrequency() {
        // Fetch the payout frequency from the repository
        return solutionProductRepository.findPayoutFrequency();
    }

    public boolean validateFSCCode(String fscCode) {
        // Validate the FSC code against predefined values
        return solutionProductRepository.isValidFSCCode(fscCode);
    }

    public double fetchSumAssured() {
        // Fetch the sum assured value from the solution product
        return solutionProductRepository.findSumAssured();
    }

    public void setPremiumTerm(int premiumTerm) {
        // Handle the business logic for setting the premium term
        fundDetailsService.setPremiumTerm(premiumTerm);
    }

    public List<String> getRenewalPaymentMethods(String frequency, String fscCode) {
        // Determine the renewal payment methods based on the frequency and FSC codes
        return solutionProductRepository.findRenewalPaymentMethods(frequency, fscCode);
    }

    public void updateBenefitTerm(int benefitTerm) {
        // Update the benefit term value in the database
        solutionProductRepository.updateBenefitTerm(benefitTerm);
    }

    public void calculateAndUpdateRiderCoverAmounts() {
        // Retrieve the records from the rider details block
        List<SolutionProduct> riderDetails = solutionProductRepository.findRiderDetails();
        for (SolutionProduct riderDetail : riderDetails) {
            if (riderDetail.getRiderCoverCode().startsWith("R")) {
                double riderCoverAmount = Math.round(riderDetail.getTotalRiderInvest() * riderDetail.getSumAssured() / (riderDetail.getSumAssured() + riderDetail.getCorrectValue().getSumAssured()) * 100.0) / 100.0;
                riderDetail.setRiderCoverAmount(riderCoverAmount);
                solutionProductRepository.saveRiderDetails(riderDetail);
            }
            if (riderDetail.getTotalRiderInvest() == 0) {
                solutionProductRepository.deleteRiderDetail(riderDetail);
            }
        }
    }

    public boolean validateProductCode(String productCode) {
        // Validate the product code
        return solutionProductRepository.validateProductCode(productCode);
    }

    public List<String> determinePaymentMethods(String frequency, String fscCode) {
        // Determine the available payment methods based on the frequency and FSC code
        return solutionProductRepository.findPaymentMethods(frequency, fscCode);
    }
}